File: 5. Sorting/06. RecursiveInsertion.py
def recursive_insertion_sort(arr, n):

    # Base case
    if n <= 1:
        return

    # Sort first n-1 elements
    recursive_insertion_sort(arr, n - 1)

    # Insert last element into the sorted part
    key = arr[n - 1]
    j = n - 2

    while j >= 0 and arr[j] > key:
        arr[j + 1] = arr[j]
        j -= 1

    arr[j + 1] = key


# Example
arr = [64, 25, 12, 22, 11]

recursive_insertion_sort(arr, len(arr))

print(arr)

File: 5. Sorting/06. RecursiveInsertion.cpp
#include <iostream>
#include <vector>
using namespace std;

void recursiveInsertionSort(vector<int>& arr, int n) {

    // Base case
    if (n <= 1)
        return;

    // Sort first n-1 elements
    recursiveInsertionSort(arr, n - 1);

    // Insert last element
    int key = arr[n - 1];
    int j = n - 2;

    while (j >= 0 && arr[j] > key) {
        arr[j + 1] = arr[j];
        j--;
    }

    arr[j + 1] = key;
}

int main() {

    vector<int> arr = {64, 25, 12, 22, 11};

    recursiveInsertionSort(arr, arr.size());

    for (int x : arr)
        cout << x << " ";

    return 0;
}